modules/apps/foundation/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/SearchPermissionCheckerImpl.java

```diff
@@ -196,6 +196,13 @@ private void _addPermissionFields(
 			roles = _resourcePermissionLocalService.getRoles(
 				companyId, className, ResourceConstants.SCOPE_INDIVIDUAL,
 				classPK, viewActionId);
+
+			roles.addAll(
+				_resourcePermissionLocalService.getRoles(
+					companyId, className,
+					ResourceConstants.SCOPE_GROUP_TEMPLATE,
+					String.valueOf(GroupConstants.DEFAULT_PARENT_GROUP_ID),
+					viewActionId));
```

Comment:
Hi guys, 

I'm not sure this is correct, but I don't know search permission checking.

I think when we index the document, it's fine to add `SCOPE_INDIVIDUAL` roles. When we change permissions we can reindex just one doc.

But is it fine to index also `ResourceConstants.SCOPE_GROUP_TEMPLATE` roles? When we change assignment on that scope then we need to reindex everything?!

On the other hand, when you index these wide-in-scope permissions, why don't you add also `SCOPE_GROUP` and `SCOPE_COMPANY` roles and move whole permission checking inside search operation?

Reply:
Hi @topolik 

Thanks for your check. Please refer to the below explanation:

1.Why index ResourceConstants.SCOPE_GROUP_TEMPLATE(or ) roles?
In search query, we use ({"terms" : {"groupRoleId" : [ "groupId-roleId"]}), and if the entry is only viewable for siteRole, own siteRole's user also can view the model.

2.Why  don't I add also SCOPE_GROUP and SCOPE_COMPANY?
Please refer to https://github.com/yuhai/liferay-portal/blob/master/modules/apps/foundation/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/SearchPermissionCheckerImpl.java#L415-L427

When compose search query, if regular role owns view permission, the query({"terms" : {"groupRoleId" : [ "groupId-roleId"]}) won't add into query. However, we can't do siteRole (as regular role) in search query because siteRole needs to use with group together.

Please refer to the fix detailed explanation from https://github.com/hhuijser/liferay-portal/pull/3030#issue-261528477

Regards,
Hai